### PR TITLE
FEAT: Create login page and auth store

### LINKED ISSUE #449: Creating the Log In page for Moderation Panel
# Problem
There is no place for the admins who look through doctor submissions to log in and get to the admin page.

# Success
A basic login page with a button for OAuth. Currently the plan is to use Google OAuth
### Requirements

- [x] Creating a page for logging in
- [x] Add a button that says log in with Google

# Implementation notes
There is no styling on figma, but if you want to look at how other login with google pages look on websites, that might be good as a model.

### LINKED ISSUE #542: FEAT: Epic - Set up Auth for our site
# Epic for Auth work
We want to allow our admins/moderators to login to the moderator dashboard. We also want to make sure someone unauthorized cannot use the moderation dashboard. 

This is work is quite large, so might be broken into separate issues for some parts. This will be the epic (parent ticket with a bunch of smaller tickets needed to complete it). 

### Requirements

- [x] Create a login page #575
- [x] Create an auth store that handles user login, logout, and keeping track of logged in status #575
- [x] Add in redirection and blocking/unblocking the moderator dashboard #575
- [x] Add support to refresh tokens #575
- [x] Integrate auth into all the requests by the client side #574
- [x] Add support for roles #571 #572 
- [x] Add auth checks to API #573 
- [x] Secure API routes to only admin and moderator roles #572
- [x] Add some tests! #570

### LINKED ISSUE #575: FEAT: auth front-end - Create a login page and add auth framework
# Problem
We want to have a way for users to login to use the moderation dashboard. 
For this issue, let's create a login page and get auth initialized!

# Success
- figure out which auth framework we want to use. Done - decided on Auth0
- initiatilize the auth framework
-  Create a login page
-  Create an auth store that handles user login, logout, and keeping track of logged in status
-  Add in redirection and blocking/unblocking the moderator dashboard
-  Add support to refresh tokens

### PR DESCRIPTION
**Pull request by ermish**
Resolves #575
Resolves #449
Part of epic #542

# 🔧 What changed
###general improvements
- ~~Fixed `$t()` and `import .. from '#app'` compiler errors! 🙌 🥳  this was by adding nuxt support to our tsconfig through the extension. Had to do some moving around of the settings to get it to work with what we already had configured.~~ these changes were moved to #560  and #566.
- improved the loading icon to have less padding
- improved the loading UI for results 
- added an error theme color (for validation error messages)

### New logged in experience
- added links to topnav and hamburger menu for when a user is logged in for convenience! 🏖️ 
- added friendly un-authorized messages to the mod dashboard if someone tries to view without permissions
- reworked the moderation dashboard access to just be based on logged in status instead of the previous env flag + route logic 

### created login page! 
- added validation
- made validation only run after typing has happened
- connected to auth store and loading state
- added redirection to moderation panel once logged in
- made a nice logging in loading indicator on the button

### created auth store
- now have a global state for checking the users logged in status
- has login function written (minus actual API call, since it doesn't exist yet)
- set up loading state events
- added roles concept and built out schema

### auth framework
- Added in supertokens logic (initializiation, supertokens ui, thirdparty login)....then ripped it all out and switched to [Auth0](https://auth0.com/). 
  - Why? supertokens didn't have great vue support and I spent several hours getting it to work. I would have gone with auth0 to start but they're expensive. I recently discovered they have NPO discounts, so it's an option again! 
- Added in auth0
  - Created initialization 
  - found a race-condition bug with their intiailization happening after vue lifecycle hooks causing weird errors. built a shared singleton to fix this (as per their recommendation) this fixed it 🐛 🔨 
  - found a nuxt bug where the initialization breaks because nuxt is try to run it at compile/ssr time. 🙄  Fixed the issue using `process.client`. This was hard to track down as the errors were not intuitive.  🐛 🔨 
  - integrated auth0 with api requests
  - Set up third party login
  - Integrated auth0 with login page and `authStore.ts` 

### moderation panel
- added check to redirect to login page if user is not logged in
- added `<Suspense>` which is a nice way to prevent the page from loading until we've confirmed they're logged in

# 🧪 Testing instructions
- try logging in (ask Phil for account) 
- ensure you can use the moderation dashboard. 
- ensure you can't use it without being logged in

# 📸 Screenshots

image

### CODE REVIEW

File: pages/moderation.vue
Comment (theyokohamalife) [inline pages/moderation.vue:14]: I know this is still WIP, but let's add this to the localization files too 🙏🏾

Comment (ermish) [reply pages/moderation.vue:14]: will do, I'm not even sure this fallback is even showing yet 😆

**Comment by theyokohamalife** [comment]
@ermish Really appreciate all the work you put into this!

Could you move the non-login and auth related changes to a separate PR? It's good practice and will make the PR easier to review. 

I think these changes should probably be moved:

> ###general improvements
> Fixed $t() and import .. from '#app' compiler errors! 🙌 🥳 this was by adding nuxt support to our tsconfig through the extension. Had to do some moving around of the settings to get it to work with what we already had configured.
> improved the loading icon to have less padding
> improved the loading UI for results
> added an error theme color (for validation error messages)

File: assets/css/tailwind.css
Comment (theyokohamalife) [inline assets/css/tailwind.css:1]: I think these are changes we should separate into another PR.

File: app.vue
Comment (theyokohamalife) [inline app.vue:1]: Most of these changes should probably be moved to another PR since they aren't directly related to the login/auth. Then, if needed, these could be merged first before the auth changes.

File: assets/icons/loading.svg
Comment (theyokohamalife) [inline assets/icons/loading.svg:1]: Let's move this as well.

File: components/Loader.vue
Comment (theyokohamalife) [inline components/Loader.vue:1]: ♻️

File: components/LocaleSelector.vue
Comment (theyokohamalife) [inline components/LocaleSelector.vue:1]: ♻️

File: components/LoginForm.vue
Comment (theyokohamalife) [inline components/LoginForm.vue:6]: We should make this i18n

File: components/ModSubmissionsEditFormFacility.vue
Comment (theyokohamalife) [inline components/ModSubmissionsEditFormFacility.vue:1]: Please move formatting changes to another PR 🙇🏾‍♀️

File: components/SearchResultsList.vue
Comment (theyokohamalife) [inline components/SearchResultsList.vue:1]: I think the changes to this file are also unrelated to login/auth

File: pages/about.vue
Comment (theyokohamalife) [inline pages/about.vue:1]: ♻️

File: pages/login.vue
Comment (theyokohamalife) [inline pages/login.vue:3]: Just curious, what's the reasoning for moving everything into a component instead of just using the page?

**Review by theyokohamalife — CHANGES_REQUESTED** [review]

**Comment by theyokohamalife** [comment]
The login page looks very different from the screenshot in your PR. I'm guessing that was scrapped and you forgot to update it. Here is what I see when I visit it:

Screenshot 2024-06-11 at 2 01 17 PM

We definitely want to go with our official name, which is "Find a Doc, Japan"

Maybe it should say something like, "Log into Find a Doc, Japan to continue to the moderation panel"

**Comment by theyokohamalife** [comment]
It seems a little odd for the login page to say "Redirecting to login page"

Maybe it should say something along the lines of, "Redirecting to Auth0 for authentication" so the user knows they're being taken to a separate platform. 

Screenshot 2024-06-11 at 2 05 48 PM

**Comment by theyokohamalife** [comment]
After I sign in with my credentials, I'm asked to authorize the app. Once I click "Accept" it takes me back to the home page at `/`
Screenshot 2024-06-11 at 2 08 22 PM

If I try to go to `/login` again, I see two copies of the top nav and a "Login!" message instead of being routed to the moderation dashboard:
Screenshot 2024-06-11 at 2 11 17 PM

**Comment by theyokohamalife** [comment]
Noticed some strange behavior after logging in. If I have my local server running, it'll redirect me to `localhost:3000/login` despite initiating the login from the Netlify preview.

And if I disable my local server, it will redirect me to `findadoc.jp/login` instead of staying with Netlify. I'm guessing something needs to be changed in the way the URL endpoints are configured on Auth0?
Screenshot 2024-06-11 at 2 08 22 PM
Screenshot 2024-06-11 at 2 19 38 PM

File: assets/css/tailwind.css
Comment (ermish) [reply assets/css/tailwind.css:1]: these are actually related to the login page fields with errors

File: assets/icons/loading.svg
Comment (ermish) [reply assets/icons/loading.svg:1]: this was required for the login button

File: components/LocaleSelector.vue
Comment (ermish) [reply components/LocaleSelector.vue:1]: I don't really want to make another PR for 4 semicolons 😓

File: components/LoginForm.vue
Comment (ermish) [reply components/LoginForm.vue:6]: done!

File: components/ModSubmissionsEditFormFacility.vue
Comment (ermish) [reply components/ModSubmissionsEditFormFacility.vue:1]: these were just from running `yarn lint --fix`. I'd have to remake my PR likely which might take as much time as 
 building the rest of auth 😿

Comment (ermish) [reply components/ModSubmissionsEditFormFacility.vue:1]: we should talk about how much we split out things. I personally think that optimizations and code refactorings that are done in the same space as work it good to be included, but work for two different features should be split. 

For example, auth and moderation dashboard feature work should be separate, but improving formatting while in a file is ok to be together.

File: components/SearchResultsList.vue
Comment (ermish) [reply components/SearchResultsList.vue:1]: these are from the svg changes I needed to make for the auth login page

File: pages/about.vue
Comment (ermish) [reply pages/about.vue:1]: also part of the auth svg changes

File: pages/login.vue
Comment (ermish) [reply pages/login.vue:3]: I was thinking we might reuse the component in the auth0 login....but now I'm not sure honestly if it's going to be in more than one place.
 I personally try to keep the pages minimal as a habit, but that's just a personal preference.

File: components/Loader.vue
Comment (ermish) [reply components/Loader.vue:1]: these are from the svg changes I needed to make for the auth login page

**Comment by ermish** [comment]
> It seems a little odd for the login page to say "Redirecting to login page"
> 
> Maybe it should say something along the lines of, "Redirecting to Auth0 for authentication" so the user knows they're being taken to a separate platform.
> 
> Screenshot 2024-06-11 at 2 05 48 PM

updated!

**Comment by ermish** [comment]
> The login page looks very different from the screenshot in your PR. I'm guessing that was scrapped and you forgot to update it. Here is what I see when I visit it:
> 
> Screenshot 2024-06-11 at 2 01 17 PM
> 
> We definitely want to go with our official name, which is "Find a Doc, Japan"
> 
> Maybe it should say something like, "Log into Find a Doc, Japan to continue to the moderation panel"

I updated the name to "Find a Doc, Japan", but I'm not sure how to change the second half yet. I wouldn't want to block merging on it, since it's not a code change. (it's in the auth0 management console)

File: app.vue
Comment (ermish) [reply app.vue:1]: split out the ones that made sense 👍

**Comment by ermish** [comment]
> Noticed some strange behavior after logging in. If I have my local server running, it'll redirect me to `localhost:3000/login` despite initiating the login from the Netlify preview.
> 
> And if I disable my local server, it will redirect me to `findadoc.jp/login` instead of staying with Netlify. I'm guessing something needs to be changed in the way the URL endpoints are configured on Auth0? Screenshot 2024-06-11 at 2 08 22 PM Screenshot 2024-06-11 at 2 19 38 PM
> 
> What I'm worried will happen is that if someone logs in from prod but has the local server running, it'll redirect to `localhost:3000` instead.

yeah, I've been trying to figure out the redirections...I think I can't fix this until I merge it so there's a login page in production (which shouldn't hurt since it's not user facing and no-one has access yet since auth0 is not connected to the backend)
I can't test in deploy preview either 😢
ok with merging to test in prod? 😬

**Comment by ermish** [comment]
@theyokohamalife ok, it's ready for another review and actually ready to merge now on my end!
I think the redirection stuff is as good as I can test without a prod `/login` link to test against.

File: components/HamburgerMenu.vue
Comment (NabbeunNabi) [inline components/HamburgerMenu.vue:138]:
```suggestion
    mounted: (el: HTMLElement) => {
```

No anyscript!

Comment (ermish) [reply components/HamburgerMenu.vue:138]: oh, this is unrelated to auth, but I'm gonna fix it!

Comment (ermish) [reply components/HamburgerMenu.vue:138]: nevermind, these types are annoying to fix, so let's move them to a separate PR to finish cleaning up types

File: cypress/e2e/moderationDashboard.cy.ts
Comment (theyokohamalife) [inline cypress/e2e/moderationDashboard.cy.ts:171]: Non-blocking: Do we need this console log here?

**Review by theyokohamalife — APPROVED** [review]
LGTM. ✨  Just need CI to pass.

File: cypress/e2e/moderationDashboard.cy.ts
Comment (ermish) [reply cypress/e2e/moderationDashboard.cy.ts]: I actually think it's helpful to see the queries we've intercepted in tests, but technically we don't need it

**Pull request merged**
